[appearance: base] Create a child pseudo render element for the ::checkmark pseudo style

shallawa opened the pull request:
#### c3100e0d365f2da44633b72b58bd637df7a0c534
<pre>
[appearance: base] Create a child pseudo render element for the ::checkmark pseudo style
<a href="https://bugs.webkit.org/show_bug.cgi?id=303606">https://bugs.webkit.org/show_bug.cgi?id=303606</a>
<a href="https://rdar.apple.com/165893004">rdar://165893004</a>

Reviewed by Antti Koivisto.

A child checkmark box will be created for ::checkmark pseudo style. It will be
created only if the appearance is `base` and the ::checkmark is not styled with
`display: none;` or `display: contents;`.

This checkmark box will be removed from the parent renderer when the appearance is
not `base` or ::checkmark is styled with `display: none;` or `display: contents;`.

The UA style sheet will have a rule to set ::checkmark style with `display: none;`
when it is unchecked. For now this rule will be added in the test style sheet.

Tests: fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html
       fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html: Added.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
(WebCore::RenderTreeBuilder::FormControls::updateCheckmark):
(WebCore::RenderTreeBuilder::FormControls::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.h:

Canonical link: <a href="https://commits.webkit.org/304390@main">https://commits.webkit.org/304390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ecd082c425ca0beb92aaaab1a229fce3663db33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135446 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46727 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143139 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1c5e03c7-c12c-46f4-a8e5-3775cdd6d346) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137315 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8461 "Hash 1ecd082c for PR 54904 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7671 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103507 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ca6d261c-28f3-4342-92ec-14eb6aceb8b9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138392 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/8461 "Hash 1ecd082c for PR 54904 does not build (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/84373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dfd2f26b-ba12-41ba-a10c-a8b4d8a5d733) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/8461 "Hash 1ecd082c for PR 54904 does not build (failure)") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3750 "Built successfully") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/8461 "Hash 1ecd082c for PR 54904 does not build (failure)") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/39588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145894 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7507 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/40159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7548 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6292 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5698 "Passed tests") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/117704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20884 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7561 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7309 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/71108 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7527 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7410 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -439,6 +439,8 @@ std::optional<Inspector::Protocol::CSS::PseudoId> InspectorCSSAgent::protocolVal
     case PseudoElementType::TargetText:
         return Inspector::Protocol::CSS::PseudoId::TargetText;
+    case PseudoElementType::Checkmark:
+        return { };
```
if you want to support this in Web Inspector right now, it should be pretty simple :)
- add `"checkmark"` to the `enum` list of `PseudoId` in `Source/JavaScriptCore/inspector/protocol/CSS.json`
- add `Checkmark: "checkmark",` to `WI.CSSManager.PseudoSelectorNames` in `Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js`
- add `case CSSManager.PseudoSelectorNames.Checkmark` to `WI.CSSManager.displayNameForPseudoId` in `Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js`

if you'd rather do that later, can you please file a bug with a FIXME comment here?

dcrousso replied:
though that assumes this will appear in the DOM tree like `::before`/`::after`.  if it doesn't, then this is correct as-is

shallawa replied:
Thanks for the fixing steps.

I think the `::checkmark` should be visible in the WebInspector since the author will have the ability to change it. So viewing it in the inspector is important. So I added your suggestions.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
+
+    for (CheckedRef child : childrenOfType<RenderElement>(flow)) {
+        if (child->style().pseudoElementType() == PseudoElementType::Checkmark) {
+            if (!inputElement->checked() || flow.style().appearance() != StyleAppearance::Base)
```
We should also be generating the checkmark when the input element isn't checked (we should remove the `!inputElement->checked()` check), it's just `visibility: hidden`.

shallawa replied:
I followed what `RenderTreeBuilder::List::updateItemMarker()` does when `style.listStyleType().isNone()` becomes true and the marker was created. It calls `m_builder.destroy(*marker);`

nt1m replied:
That part is correct, I'm just saying, we should remove the `!inputElement->checked()` check and keep the `flow.style().appearance() != StyleAppearance::Base` check.

So the resulting code would be:

```
if (flow.style().appearance() != StyleAppearance::Base)
    m_builder.destroy(child);
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
+        }
+    }
+
+    if (!inputElement->checked() || flow.style().appearance() != StyleAppearance::Base)
```
ditto

shallawa replied:
I think creating the checkmark when 
```
 if (inputElement->checked() && flow.style().appearance() == StyleAppearance::Base)
```
is false will affect SP3. I think we should create the checkmark only if the above condition is true.

But if the checkmark is already created and the above condition becomes false, we may destroy the checkmark or hide it.

nt1m replied:
We should still gate the checkmark behind `flow.style().appearance() == StyleAppearance::Base`, which shouldn't affect SP3, since SP3 isn't using `appearance: base`

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
     return *innerRenderer;
 }
 
+void RenderTreeBuilder::FormControls::updateCheckmark(RenderBlockFlow& flow)
```
I think this function should also handle destroying `::checkmark` when the `::checkmark` style is `display: none` (`::marker` doesn't need to handle this because you can't set `display: none` on `::marker` at a style level).

nt1m replied:
One testcase of this is:
```
<style>
::checkmark {
  ... styles for checkmark
}
.remove-checkmark::checkmark {
  display: none;
}
</style>
<input type=checkbox id=checkbox>
<script>
onload = () => {
  checkbox.classList.add("remove-checkmark");
}
</script>
```

at the same places you check `flow.style().appearance() != StyleAppearance::Base` , you can also check if the pseudo style is display: none. (RenderTreeUpdaterViewTransition does that)

shallawa replied:
I think `::checkmark` should follow the same pattern of `::marker` to be consistent with other pseudo styles.
```
li::marker {
  ...  
  content: none; /* This is not respected in WebKit but it respected in other browsers. */
  display: none; /* This is not respected in all browsers. */
  visibility: hidden; /* This is not respected in all browsers. */
}
```
For hiding the `::marker` renderer one should use `content: none;`. Similarly for hiding the `::checkmark` renderer, the `checked` attribute should be removed or set to false.

Having different ways to hide the `::checkmark` renderer is confusing. And making hiding the `::checkmark` renderer different from other pseudo renders like `::marker` is another source of confusion.

nt1m replied:
::marker is the odd one out here because it's inline content. all the pseudo-elements (::before/::after/::view-transition-*/::backdrop/etc.) have multiple ways to hide, this is what's defined in the standard. The diff I sent offline should cover all the different cases.

shallawa replied:
This diff does not work.

I suggest landing this PR as is. And we can file a bug for the HTMLInputElement not updating its pseudo element since this is a separate issue.

shallawa replied:
Also I do not understand why ::marker is different. And where in the specs we say display: none; should be respected or ignored?

shallawa replied:
> ::marker is the odd one out here because it's inline content

what does `inline content` mean in this context? A renderer should be created for `::checkmark` or `::marker`. I do not see a difference here. Only the style for each of them should be different.

nt1m replied:
::marker has a special set of properties that work on it: https://drafts.csswg.org/css-lists-3/#marker-properties

shallawa replied:
> This diff does not work.

I take this back. I was wrong. I had other changes on my machine while trying your diff and  it did not work. I cleaned everything and I applied the diff and it worked.

For reference I attached the diff here. Basically we need to call `invalidateStyleInternal();` at the end of `HTMLInputElement::setChecked()`. And we need to check `pseudoStyle->display() == DisplayType::None` instead of checking `!inputElement->checked()`.

[setChecked-invalidateStyleInternal.patch](https://github.com/user-attachments/files/24117026/setChecked-invalidateStyleInternal.patch)

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
+        }
+    }
+
+    if (!inputElement->checked() || flow.style().appearance() != StyleAppearance::Base)
```
The `!inputElement->checked()` checks appear to be back

anttijk commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
+
+    for (CheckedRef child : childrenOfType<RenderElement>(flow)) {
+        if (child->style().pseudoElementType() == PseudoElementType::Checkmark) {
+            if (flow.style().appearance() != StyleAppearance::Base || pseudoStyle->display() == DisplayType::None)
```
You could unify this style check and the one below to guarantee they stay in sync (as a lambda for example).

anttijk commented on Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
```diff
@@ -99,4 +103,47 @@ RenderBlock& RenderTreeBuilder::FormControls::findOrCreateParentForChild(RenderM
+    }
+
+    if (flow.style().appearance() != StyleAppearance::Base || pseudoStyle->display() == DisplayType::None)
+        return;
```
`DisplayType::Contents` shouldn't create a box either. Are all other display property values ok?

anttijk replied:
Actually `display:contents` gets changed to `none` in style adjuster so that comment can be ignored.

nt1m reviewed (commented):
LGTM too! Would be good to add a test exercising display: none dynamic mutations on ::checkmark

shallawa committed:
[appearance: base] Create a child pseudo render element for the ::checkmark pseudo style
https://bugs.webkit.org/show_bug.cgi?id=303606
rdar://165893004

Reviewed by Antti Koivisto.

A child checkmark box will be created for ::checkmark pseudo style. It will be
created only if the appearance is `base` and the ::checkmark is not styled with
`display: none;` or `display: contents;`.

This checkmark box will be removed from the parent renderer when the appearance is
not `base` or ::checkmark is styled with `display: none;` or `display: contents;`.

The UA style sheet will have a rule to set ::checkmark style with `display: none;`
when it is unchecked. For now this rule will be added in the test style sheet.

Tests: fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html
       fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html: Added.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.cpp:
(WebCore::RenderTreeBuilder::FormControls::updateCheckmark):
(WebCore::RenderTreeBuilder::FormControls::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderFormControls.h:

Canonical link: https://commits.webkit.org/304390@main
